File: 31. Next Permutation.py
import sys
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        ind1, ind2 = len(nums)-1, len(nums)-1
        for i in xrange(len(nums)-1, -1, -1):
            if nums[i] > nums[i-1]:
                ind1 = i - 1
                break
        minn = sys.maxint
        for i in xrange(ind1+1, len(nums)):
            if nums[i] < minn and nums[i] > nums[ind1]:
                minn = nums[i]
                ind2 = i
        nums[ind1], nums[ind2] = nums[ind2], nums[ind1]
        temp = sorted(nums[ind1+1:])
        for i in xrange(ind1+1, len(nums)):
            nums[i] = temp[i-ind1-1]

File: 31. Next Permutation.cpp
class Solution {
public:
    void swap(int& a, int& b)
    {
        int t = a;
        a = b;
        b = t;
    }
    void nextPermutation(vector<int>& nums) {
        int ind1 = nums.size()-1, ind2 = nums.size()-1;
        for (int i = nums.size()-1; i >= 0; i--)
        {
            if (nums[i] > nums[i-1])
            {
                ind1 = i-1;
                break;
            }
        }
        int min = INT_MAX;
        for (int i = ind1+1; i < nums.size(); i++)
        {
            if (nums[i] < min && nums[i] > nums[ind1])
            {
                min = nums[i];
                ind2 = i;
            }
        }
        if (ind1 == -1)
            sort(nums.begin(), nums.end());
        else
        {
            swap(nums[ind1], nums[ind2]);
            vector <int> temp (nums.begin() + ind1 + 1, nums.end());
            sort(temp.begin(), temp.end());
            for (int i = ind1 + 1; i < nums.size(); i++)
                nums[i] = temp[i-ind1-1];
        }
    }
};
